satellite/satellitedb/usercredits.go

```diff
@@ -0,0 +1,172 @@
+func (c *usercredits) GetAvailableCredits(ctx context.Context, referrerID uuid.UUID, expirationEndDate time.Time) ([]console.UserCredit, error) {
+	availableCredits, err := c.db.All_UserCredit_By_UserId_And_ExpiresAt_Greater_And_CreditsUsedInCents_Less_CreditsEarnedInCents_OrderBy_Asc_ExpiresAt(ctx,
+		dbx.UserCredit_UserId(referrerID[:]),
+		dbx.UserCredit_ExpiresAt(expirationEndDate),
+	)
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+
+	return fromDBX(availableCredits)
+}
+
+// Create insert a new record of user credit
+func (c *usercredits) Create(ctx context.Context, userCredit console.UserCredit) (*console.UserCredit, error) {
+	createdUser, err := c.db.Create_UserCredit(ctx,
+		dbx.UserCredit_UserId(userCredit.UserID[:]),
+		dbx.UserCredit_OfferId(userCredit.OfferID),
+		dbx.UserCredit_CreditsEarnedInCents(userCredit.CreditsEarnedInCents),
+		dbx.UserCredit_ExpiresAt(userCredit.ExpiresAt),
+		dbx.UserCredit_Create_Fields{
+			ReferredBy: dbx.UserCredit_ReferredBy(userCredit.ReferredBy[:]),
+		},
+	)
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+
+	return convertDBX(createdUser)
+}
+
+// UpdateAvailableCredits updates user's available credits based on their spending and the time of their spending
+func (c *usercredits) UpdateAvailableCredits(ctx context.Context, creditsToCharge int, id uuid.UUID, expirationEndDate time.Time) (remainingCharge int, err error) {
+	tx, err := c.db.Open(ctx)
+	if err != nil {
+		return creditsToCharge, errs.Wrap(err)
+	}
+
+	availableCredits, err := tx.All_UserCredit_By_UserId_And_ExpiresAt_Greater_And_CreditsUsedInCents_Less_CreditsEarnedInCents_OrderBy_Asc_ExpiresAt(ctx,
+		dbx.UserCredit_UserId(id[:]),
+		dbx.UserCredit_ExpiresAt(expirationEndDate),
+	)
+	if err != nil {
+		return creditsToCharge, errs.Wrap(errs.Combine(err, tx.Rollback()))
+	}
+	if len(availableCredits) == 0 {
+		return creditsToCharge, errs.Combine(errs.New("No available credits"), tx.Commit())
+	}
+
+	var infos []updateInfo
+	remainingCharge = creditsToCharge
+	for _, credit := range availableCredits {
+		if remainingCharge == 0 {
+			break
+		}
+
+		creditsForUpdate := credit.CreditsEarnedInCents - credit.CreditsUsedInCents
+
+		if remainingCharge < creditsForUpdate {
+			creditsForUpdate = remainingCharge
+		}
+
+		infos = append(infos, updateInfo{
+			id:      credit.Id,
+			credits: creditsForUpdate,
+		})
+
+		remainingCharge -= creditsForUpdate
+	}
+
+	statement := `UPDATE user_credits SET
+		credits_used_in_cents = CASE id ` + convertToSQLFormat(infos)
+
+	_, err = tx.Tx.ExecContext(ctx, c.db.Rebind(statement))
+	if err != nil {
+		return creditsToCharge, errs.Wrap(errs.Combine(err, tx.Rollback()))
+	}
+	return remainingCharge, errs.Wrap(tx.Commit())
+}
+
+func convertToSQLFormat(updateInfo []updateInfo) (updateStr string) {
+	for i, info := range updateInfo {
+		updateStr += `WHEN ` + strconv.Itoa(info.id) + ` THEN ` + strconv.Itoa(info.credits)
```

Comment:
No string addition of values for sql queries, it's easy to get the escaping wrong, use statement arguments. While it's probably fine here, I don't want to have a situation where somebody modifies this statement and accidentally adds something that causes a SQL injection.

Reply:
See https://www.owasp.org/index.php/SQL_Injection for more information.

Reply:
I made the change